File: my-frontend/src/services/quizService.ts
import axiosInstance from "./axiosInstance";
import { PageResponse } from "@/interfaces";
import {
    addQuestionsBulk,
    QuestionRequestDTO,
    OptionRequestDTO,
    mapAiListToCreatePayload,
    QuestionCreatePayload,
} from "./questionService";
import { handleApiError } from "@/utils/apiErrorHandler";

/* =========================
 * Types & Interfaces
 * ========================= */

export enum Visibility {
    PUBLIC = "PUBLIC",
    UNLISTED = "UNLISTED",
    PASSWORD = "PASSWORD",
    PRIVATE = "PRIVATE",
}

export enum Difficulty {
    EASY = "EASY",
    MEDIUM = "MEDIUM",
    HARD = "HARD",
    EXPERT = "EXPERT",
}

export interface QuizCreateRequest {
    title: string;
    description?: string;
    coverImageUrl?: string;
    tags?: string[];
    difficulty?: Difficulty;
    estimatedMinutes?: number;
    visibility?: Visibility;
    accessPassword?: string;
    allowedUserIds?: string[];
}

export interface QuizUpdateRequest {
    title: string;
    description?: string;
    coverImageUrl?: string;
    tags?: string[];
    difficulty?: Difficulty;
    estimatedMinutes?: number;
    visibility?: Visibility;
    accessPassword?: string;
    allowedUserIds?: string[];
}

export interface QuizDetailResponse {
    quizId: string;
    title: string;
    description: string;
    slug: string;
    coverImageUrl?: string;
    tags: string[];
    difficulty: Difficulty;
    estimatedMinutes: number;
    visibility: Visibility;
    published: boolean;
    creator: {
        userId: string;
        name: string;
        email: string;
    };
    isOwner: boolean;
    canPlay: boolean;
    totalQuestions: number;
    averageScore: number;
    viewCount: number;
    startCount: number;
    completionCount: number;
    createdAt: string;
    updatedAt: string;
}

export interface QuizSummaryDto {
    quizId: string;
    title: string;
    description: string;
    slug: string;
    coverImageUrl?: string;
    difficulty: Difficulty;
    estimatedMinutes: number;
    totalQuestions: number;
    averageScore: number;
    viewCount: number;
    startCount: number;
    completionCount: number;
    creator: {
        userId: string;
        name: string;
        avatar?: string;
    };
    createdAt: string;
}

/* =========================
 * PUBLIC ENDPOINTS - Browse Quizzes
 * ========================= */

/**
 * Get all published public quizzes with optional search
 */
export const getPublishedQuizzes = async (
    keyword?: string,
    page: number = 0,
    size: number = 20,
    sortBy: string = "createdAt",
    sortDirection: "ASC" | "DESC" = "DESC"
): Promise<PageResponse<QuizSummaryDto>> => {
    try {
        const params = new URLSearchParams();
        if (keyword) params.append("keyword", keyword);
        params.append("page", page.toString());
        params.append("size", size.toString());
        params.append("sortBy", sortBy);
        params.append("sortDirection", sortDirection);

        const response = await axiosInstance.get(`/quizzes/published`, {
            params,
        });
        return response.data;
    } catch (error) {
        handleApiError(error, "Failed to fetch published quizzes");
        throw error;
    }
};

/**
 * Get quiz by slug (public access with optional password)
 */
export const getQuizBySlug = async (
    slug: string,
    password?: string
): Promise<QuizDetailResponse> => {
    try {
        const params = new URLSearchParams();
        if (password) params.append("password", password);

        const response = await axiosInstance.get(`/quizzes/slug/${slug}`, {
            params,
        });
        return response.data;
    } catch (error) {
        handleApiError(error, "Failed to fetch quiz by slug");
        throw error;
    }
};

/**
 * Get trending quizzes sorted by popularity
 */
export const getTrendingQuizzes = async (
    page: number = 0,
    size: number = 20
): Promise<PageResponse<QuizSummaryDto>> => {
    try {
        const response = await axiosInstance.get(`/quizzes/trending`, {
            params: { page, size },
        });
        return response.data;
    } catch (error) {
        handleApiError(error, "Failed to fetch trending quizzes");
        throw error;
    }
};

/**
 * Get quizzes by difficulty level
 */
export const getQuizzesByDifficulty = async (
    difficulty: Difficulty,
    page: number = 0,
    size: number = 20
): Promise<PageResponse<QuizSummaryDto>> => {
    try {
        const response = await axiosInstance.get(
            `/quizzes/difficulty/${difficulty}`,
            { params: { page, size } }
        );
        return response.data;
    } catch (error) {
        handleApiError(error, `Failed to fetch ${difficulty} quizzes`);
        throw error;
    }
};

/**
 * Get quizzes by category
 */
export const getQuizzesByCategory = async (
    categoryId: number,
    page: number = 0,
    size: number = 20
): Promise<PageResponse<QuizSummaryDto>> => {
    try {
        const response = await axiosInstance.get(
            `/quizzes/category/${categoryId}`,
            { params: { page, size } }
        );
        return response.data;
    } catch (error) {
        handleApiError(error, "Failed to fetch quizzes by category");
        throw error;
    }
};

/**
 * Get popular quizzes in a specific category
 */
export const getPopularByCategory = async (
    categoryId: number,
    page: number = 0,
    size: number = 20
): Promise<PageResponse<QuizSummaryDto>> => {
    try {
        const response = await axiosInstance.get(
            `/quizzes/category/${categoryId}/popular`,
            { params: { page, size } }
        );
        return response.data;
    } catch (error) {
        handleApiError(error, "Failed to fetch popular quizzes");
        throw error;
    }
};

/**
 * Get all published quizzes by user
 */
export const getQuizzesByUser = async (
    userId: string,
    page: number = 0,
    size: number = 20
): Promise<PageResponse<QuizSummaryDto>> => {
    try {
        const response = await axiosInstance.get(`/quizzes/users/${userId}`, {
            params: { page, size },
        });
        return response.data;
    } catch (error) {
        handleApiError(error, "Failed to fetch user quizzes");
        throw error;
    }
};

/**
 * Advanced search with multiple filters
 */
export const searchQuizzes = async (
    keyword?: string,
    difficulty?: Difficulty,
    categoryId?: number,
    page: number = 0,
    size: number = 20
): Promise<PageResponse<QuizSummaryDto>> => {
    try {
        const params = new URLSearchParams();
        if (keyword) params.append("keyword", keyword);
        if (difficulty) params.append("difficulty", difficulty);
        if (categoryId) params.append("categoryId", categoryId.toString());
        params.append("page", page.toString());
        params.append("size", size.toString());

        const response = await axiosInstance.get(`/quizzes/search`, {
            params,
        });
        return response.data;
    } catch (error) {
        handleApiError(error, "Failed to search quizzes");
        throw error;
    }
};

/* =========================
 * AUTHENTICATED ENDPOINTS - Quiz Management
 * ========================= */

/**
 * Get quiz by ID (requires auth for private quizzes)
 */
export const getQuizById = async (
    quizId: string
): Promise<QuizDetailResponse> => {
    try {
        const response = await axiosInstance.get(`/quizzes/${quizId}`);
        return response.data;
    } catch (error) {
        handleApiError(error, "Failed to fetch quiz");
        throw error;
    }
};

/**
 * Get all quizzes created by current user (published + unpublished)
 */
export const getMyQuizzes = async (
    page: number = 0,
    size: number = 20
): Promise<PageResponse<QuizSummaryDto>> => {
    try {
        const response = await axiosInstance.get(`/quizzes/my-quizzes`, {
            params: { page, size },
        });
        return response.data;
    } catch (error) {
        handleApiError(error, "Failed to fetch your quizzes");
        throw error;
    }
};

/**
 * Get draft quizzes (unpublished only)
 */
export const getMyDrafts = async (
    page: number = 0,
    size: number = 20
): Promise<PageResponse<QuizSummaryDto>> => {
    try {
        const response = await axiosInstance.get(`/quizzes/my-drafts`, {
            params: { page, size },
        });
        return response.data;
    } catch (error) {
        handleApiError(error, "Failed to fetch draft quizzes");
        throw error;
    }
};

/**
 * Create a new quiz
 */
export const createQuiz = async (
    request: QuizCreateRequest
): Promise<QuizDetailResponse> => {
    try {
        const response = await axiosInstance.post(`/quizzes`, request);
        return response.data;
    } catch (error) {
        handleApiError(error, "Failed to create quiz");
        throw error;
    }
};

/**
 * Update quiz details
 */
export const updateQuiz = async (
    quizId: string,
    request: QuizUpdateRequest
): Promise<QuizDetailResponse> => {
    try {
        const response = await axiosInstance.put(`/quizzes/${quizId}`, request);
        return response.data;
    } catch (error) {
        handleApiError(error, "Failed to update quiz");
        throw error;
    }
};

/**
 * Publish quiz to make it publicly available
 */
export const publishQuiz = async (quizId: string): Promise<void> => {
    try {
        await axiosInstance.post(`/quizzes/${quizId}/publish`);
    } catch (error) {
        handleApiError(error, "Failed to publish quiz");
        throw error;
    }
};

/**
 * Unpublish quiz to make it private/draft
 */
export const unpublishQuiz = async (quizId: string): Promise<void> => {
    try {
        await axiosInstance.post(`/quizzes/${quizId}/unpublish`);
    } catch (error) {
        handleApiError(error, "Failed to unpublish quiz");
        throw error;
    }
};

/**
 * Soft delete quiz (can be restored)
 */
export const softDeleteQuiz = async (quizId: string): Promise<void> => {
    try {
        await axiosInstance.delete(`/quizzes/${quizId}`);
    } catch (error) {
        handleApiError(error, "Failed to delete quiz");
        throw error;
    }
};

/**
 * Hard delete quiz permanently
 */
export const hardDeleteQuiz = async (quizId: string): Promise<void> => {
    try {
        await axiosInstance.delete(`/quizzes/${quizId}/hard`);
    } catch (error) {
        handleApiError(error, "Failed to permanently delete quiz");
        throw error;
    }
};

/**
 * Restore a soft-deleted quiz
 */
export const restoreQuiz = async (quizId: string): Promise<void> => {
    try {
        await axiosInstance.post(`/quizzes/${quizId}/restore`);
    } catch (error) {
        handleApiError(error, "Failed to restore quiz");
        throw error;
    }
};

/**
 * Duplicate quiz with all questions
 */
export const duplicateQuiz = async (
    quizId: string
): Promise<QuizDetailResponse> => {
    try {
        const response = await axiosInstance.post(
            `/quizzes/${quizId}/duplicate`
        );
        return response.data;
    } catch (error) {
        handleApiError(error, "Failed to duplicate quiz");
        throw error;
    }
};

/* =========================
 * ANALYTICS & TRACKING
 * ========================= */

/**
 * Increment view count for quiz
 */
export const incrementViewCount = async (quizId: string): Promise<void> => {
    try {
        await axiosInstance.post(`/quizzes/${quizId}/increment-view`);
    } catch (error) {
        console.debug("Failed to increment view count", error);
    }
};

/**
 * Increment start count for quiz
 */
export const incrementStartCount = async (quizId: string): Promise<void> => {
    try {
        await axiosInstance.post(`/quizzes/${quizId}/increment-start`);
    } catch (error) {
        console.debug("Failed to increment start count", error);
    }
};

/**
 * Increment completion count for quiz
 */
export const incrementCompletionCount = async (
    quizId: string
): Promise<void> => {
    try {
        await axiosInstance.post(`/quizzes/${quizId}/increment-completion`);
    } catch (error) {
        console.debug("Failed to increment completion count", error);
    }
};

File: QuizAppBackend/src/main/java/com/kkunquizapp/QuizAppBackend/quiz/service/QuizService.java
package com.kkunquizapp.QuizAppBackend.quiz.service;

import com.kkunquizapp.QuizAppBackend.quiz.dto.*;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

/**
 * Quiz Service Interface
 * Comprehensive quiz management operations
 */
public interface QuizService {

    // ==================== CREATE ====================
    /**
     * Create a new quiz
     * @param request Quiz creation request
     * @param creatorId ID of quiz creator
     * @return Created quiz details
     */
    QuizDetailResponse createQuiz(QuizCreateRequest request, UUID creatorId);

    // ==================== READ ====================
    /**
     * Get quiz by slug
     * @param slug Quiz slug (URL-friendly identifier)
     * @param userId User ID (nullable for public access)
     * @param password Password if quiz is password-protected
     * @return Quiz details with questions
     */
    QuizDetailResponse getQuizDetailBySlug(String slug, UUID userId, String password);

    /**
     * Get quiz by ID
     * @param quizId Quiz ID
     * @param userId User ID
     * @return Quiz details with questions
     */
    QuizDetailResponse getQuizDetailById(UUID quizId, UUID userId);

    /**
     * Get all published quizzes with optional keyword search
     * @param keyword Optional search keyword
     * @param pageable Pagination info
     * @return Page of quiz summaries
     */
    Page<QuizSummaryDto> getPublicQuizzes(String keyword, Pageable pageable);

    /**
     * Get all quizzes created by user
     * @param userId Creator ID
     * @param pageable Pagination info
     * @return Page of quiz summaries
     */
    Page<QuizSummaryDto> getMyQuizzes(UUID userId, Pageable pageable);

    /**
     * Get all draft quizzes (unpublished) for user
     * @param userId Creator ID
     * @param pageable Pagination info
     * @return Page of draft quiz summaries
     */
    Page<QuizSummaryDto> getMyDrafts(UUID userId, Pageable pageable);

    /**
     * Get trending quizzes (sorted by views/plays)
     * @param pageable Pagination info
     * @return Page of trending quiz summaries
     */
    Page<QuizSummaryDto> getTrendingQuizzes(Pageable pageable);

    /**
     * Get quizzes by difficulty level
     * @param difficulty EASY, MEDIUM, or HARD
     * @param pageable Pagination info
     * @return Page of quiz summaries
     */
    Page<QuizSummaryDto> getQuizzesByDifficulty(String difficulty, Pageable pageable);

    /**
     * Get quizzes by category
     * @param categoryId Category ID
     * @param pageable Pagination info
     * @return Page of quiz summaries
     */
    Page<QuizSummaryDto> getQuizzesByCategory(Integer categoryId, Pageable pageable);

    /**
     * Advanced search with multiple filters
     * @param keyword Search keyword
     * @param difficulty Filter by difficulty
     * @param categoryId Filter by category
     * @param pageable Pagination info
     * @return Page of matching quiz summaries
     */
    Page<QuizSummaryDto> searchQuizzes(String keyword, String difficulty, Integer categoryId, Pageable pageable);

    /**
     * Get popular quizzes filtered by category
     * Sorted by view count and play count
     * @param categoryId Category ID
     * @param pageable Pagination info
     * @return Page of popular quiz summaries
     */
    Page<QuizSummaryDto> getPopularByCategory(Integer categoryId, Pageable pageable);

    // ==================== UPDATE ====================
    /**
     * Update quiz details
     * @param quizId Quiz ID
     * @param request Update request
     * @param userId User ID (must be owner)
     * @return Updated quiz details
     */
    QuizDetailResponse updateQuiz(UUID quizId, QuizUpdateRequest request, UUID userId);

    // ==================== PUBLISH ====================
    /**
     * Publish a quiz (make it publicly available)
     * Quiz must have at least 3 questions to publish
     * @param quizId Quiz ID
     * @param userId User ID (must be owner)
     */
    void publishQuiz(UUID quizId, UUID userId);

    /**
     * Unpublish a quiz (make it private/draft)
     * @param quizId Quiz ID
     * @param userId User ID (must be owner)
     */
    void unpublishQuiz(UUID quizId, UUID userId);

    // ==================== DELETE ====================
    /**
     * Soft delete a quiz (mark as deleted but keep data)
     * @param quizId Quiz ID
     * @param userId User ID (must be owner)
     */
    void softDeleteQuiz(UUID quizId, UUID userId);

    /**
     * Hard delete a quiz (permanent removal)
     * All questions and options are also deleted
     * @param quizId Quiz ID
     * @param userId User ID (must be owner)
     */
    void hardDeleteQuiz(UUID quizId, UUID userId);

    /**
     * Restore a soft-deleted quiz
     * @param quizId Quiz ID
     * @param userId User ID (must be owner)
     */
    void restoreQuiz(UUID quizId, UUID userId);

    // ==================== DUPLICATE ====================
    /**
     * Duplicate a quiz with all its questions and options
     * New quiz is created as PRIVATE draft for the user
     * @param quizId Source quiz ID
     * @param userId User ID (owner of duplicate)
     * @return New quiz details
     */
    QuizDetailResponse duplicateQuiz(UUID quizId, UUID userId);

    // ==================== ANALYTICS & TRACKING ====================
    /**
     * Increment play count (when quiz is played/attempted)
     * @param quizId Quiz ID
     */
    void incrementPlayCount(UUID quizId);

    /**
     * Increment view count (when quiz is viewed)
     * @param quizId Quiz ID
     */
    void incrementViewCount(UUID quizId);

    /**
     * Increment start count (when user starts quiz)
     * @param quizId Quiz ID
     */
    void incrementStartCount(UUID quizId);

    /**
     * Increment completion count (when user completes quiz)
     * @param quizId Quiz ID
     */
    void incrementCompletionCount(UUID quizId);

    /**
     * Update average score
     * Called after each quiz completion to recalculate average
     * @param quizId Quiz ID
     * @param score New score to average in
     */
    void updateAverageScore(UUID quizId, double score);
}